bcda/cclf/csv.go

```diff
@@ -0,0 +1,212 @@
+// table will have it's import status updated.
+func (importer CSVImporter) ProcessCSV(csv csvFile) error {
+	ctx := context.Background()
+	repository := postgres.NewRepository(importer.Database)
+	exists, err := repository.GetCCLFFileExistsByName(ctx, csv.metadata.name)
+	if err != nil {
+		err = fmt.Errorf("database query returned an error: %s", err)
+		return err
+	}
+	if exists {
+		return &ers.AttributionFileAlreadyExists{Filename: csv.metadata.name}
+	}
+
+	importer.Logger.Infof("Importing CSV file %s...", csv.metadata.name)
+	conn, err := stdlib.AcquireConn(importer.Database)
+	if err != nil {
+		return err
+	}
+
+	defer utils.CloseAndLog(logrus.WarnLevel, func() error { return stdlib.ReleaseConn(importer.Database, conn) })
+
+	tx, err := conn.BeginEx(ctx, nil)
+	if err != nil {
+		err = fmt.Errorf("failed to start transaction: %w", err)
+		return err
+	}
+
+	rtx := postgres.NewRepositoryPgxTx(tx)
+
+	defer func() {
+		if err != nil {
+			if err1 := tx.Rollback(); err1 != nil {
+				importer.Logger.Errorf("Failed to rollback transaction: %s", err.Error())
+			}
+			return
+		}
+	}()
+
+	// CCLF model corresponds with a database record
+	record := models.CCLFFile{
+		CCLFNum:         csv.metadata.cclfNum,
+		Name:            csv.metadata.name,
+		ACOCMSID:        csv.metadata.acoID,
+		Timestamp:       csv.metadata.timestamp,
+		PerformanceYear: csv.metadata.perfYear,
+		ImportStatus:    constants.ImportInprog,
+		Type:            csv.metadata.fileType,
+	}
+
+	record.ID, err = rtx.CreateCCLFFile(ctx, record)
+	if err != nil {
+		err := fmt.Errorf("database error when calling CreateCCLFFile(): %s", err)
+		return err
+	}
+
+	csv.metadata.fileID = record.ID
+
+	rows, count, err := importer.prepareCSVData(csv.data, record.ID)
+	if err != nil {
+		return err
+	}
+
+	num, err := tx.CopyFrom(pgx.Identifier{"cclf_beneficiaries"}, []string{"file_id", "mbi"}, pgx.CopyFromRows(rows))
+	if count != num {
+		return fmt.Errorf("Unexpected number of records imported (expected: %d, actual: %d)", count, num)
+	}
+
+	err = rtx.UpdateCCLFFileImportStatus(ctx, csv.metadata.fileID, constants.ImportComplete)
+	if err != nil {
+		return fmt.Errorf("database error when calling UpdateCCLFFileImportStatus(): %s", csv.metadata.name)
+	}
+
+	if err = tx.Commit(); err != nil {
+		return fmt.Errorf("failed to commit database transaction: %s", err)
+	}
+
+	successMsg := fmt.Sprintf("Successfully imported %d records from csv file %s.", num, csv.metadata.name)
+	importer.Logger.WithFields(logrus.Fields{"imported_count": num}).Info(successMsg)
+	return nil
+}
+
+func (importer CSVImporter) prepareCSVData(csvfile *bytes.Reader, id uint) ([][]interface{}, int, error) {
+	var rows [][]interface{}
+	r := csv.NewReader(csvfile)
+
+	_, err := r.Read()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	count := 0
+
+	for {
+		var record []string
+		record, err = r.Read()
+		if err == io.EOF {
+			err = nil
```

Comment:
I set this error to return nil here because EOF is an error that we expect to have when we read to the end of the file; I didn't want to bubble that error up the stack or handle it in the caller. By setting it nil here, it saves an an error handling step that we don't really need, since we want to reach EOF.